### PR TITLE
[mlir][vector] Fix 0-d vector transfer mask inference

### PR DESCRIPTION
**Pull request by dcaballe**
When we infer the mask of a transfer operation which results in a single `i1` element, we can use `vector<i1>` or `vector<1xi1>` to represent it. To avoid issues with this type mismatch, this PR fixes the mask inference logic to always generate `vector<1xi1>` for these cases. We can enable 0-d masks if they are eventually needed.

See: https://github.com/llvm/llvm-project/issues/116197

### CODE REVIEW

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp [hunk 1]
```diff
@@ -4122,6 +4122,10 @@ VectorType mlir::vector::inferTransferOpMaskType(VectorType vecType,
   assert(invPermMap && "Inversed permutation map couldn't be computed");
   SmallVector<int64_t, 8> maskShape = invPermMap.compose(vecType.getShape());
 
+  // Turn a 0-D mask into a single-element 1-D mask.
```
Comment (banach-space) [inline mlir/lib/Dialect/Vector/IR/VectorOps.cpp:4125]: [nit] Personally, I'd appreciate a note saying that ATM `vector.mask` does not support 0-D masks (enforced [here](https://github.com/llvm/llvm-project/blob/4b50ec43d03d9ba9b43edd9a4743951f6498b964/mlir/include/mlir/Dialect/Vector/IR/VectorOps.td#L2509-L2510)). And that's basically the reason to "upgrade" this mask (which to me is a very good reason)

File: mlir/test/Dialect/Vector/invalid.mlir [hunk 2]
```diff
@@ -1752,6 +1752,22 @@ func.func @vector_mask_non_maskable_op(%a : vector<3x4xf32>) -> vector<3x4xf32>
 
 // -----
 
+// We can support 0-D masks if eventually needed.
```
Comment (banach-space) [inline mlir/test/Dialect/Vector/invalid.mlir:1755]: [nit] Why not add this comment where this restriction is enforced? https://github.com/llvm/llvm-project/blob/4b50ec43d03d9ba9b43edd9a4743951f6498b964/mlir/include/mlir/Dialect/Vector/IR/VectorOps.td#L2509-L2510 

And to the Op docs :)

**Review by banach-space — APPROVED** [review]
LGTM, thanks for the fix!

**Review by Groverkss — DISMISSED** [review]
I don't really understand why we are forcing a vector<1xi1> mask instead of `vector<i1>` mask. Something seems wrong here.

I'll take 3 examples:

```
vector.transfer_read %tensor[], permutation_map<() -> (0, 0, 0)> : tensor<f32>, vector<4x4x4xf32>
vector.transfer_read %tensor[%idx], permutation_map<(d0) -> (0, 0, 0)> : tensor<1xf32>, vector<4x4x4xf32>
vector.transfer_read %tensor[%idx, %idx2], permutation_map<(d0, d1) -> (0, 0, 0)> : tensor<1x1xf32>, vector<4x4x4xf32>
```

From this PR, the mask for all of these vector.transfer_read will be vector<1xi1>, which doesn't really make sense to me. The permutation_map attribute specifies a mapping from the memory/tensor space to vector space. The masking needs to be done on the memory/tensor space. Each of these examples has a different dimensionality for the memory/tensor space. I would expect the masks to be:

```
vector.transfer_read %tensor[], permutation_map<() -> (0, 0, 0)> : tensor<f32>, vector<4x4x4xf32> // mask: vector<f32>
vector.transfer_read %tensor[%idx], permutation_map<(d0) -> (0, 0, 0)> : tensor<1xf32>, vector<4x4x4xf32> // mask: vector<1xf32>
vector.transfer_read %tensor[%idx, %idx2], permutation_map<(d0, d1) -> (0, 0, 0)> : tensor<1x1xf32>, vector<4x4x4xf32> // mask: vector<1x1xf32>
```

Based on the definition, the mask dimensionality should always match the dimensionality of the domain.

Instead of forcing the mask to be always vector<1xf32>, can we fix the inference to infer a mask of dimensionality as domain of the permutation map, and any dimension not used in the result of the permutation_map is simply 1, since it's broadcasted. This will also make the operation much more consistent w.r.t. masking. You can always expect the dimensionality of the mask to be same as the memory/tensor.

**Comment by banach-space** [comment]
> I don't really understand why we are forcing a vector<1xi1> mask instead of vector mask. Something seems wrong here.

I’d be cautious about concluding that something is broken - at least not yet! 😊

> I'll take 3 examples:
> ```
> vector.transfer_read %tensor[], permutation_map<() -> (0, 0, 0)> : tensor<f32>, vector<4x4x4xf32>
> vector.transfer_read %tensor[%idx], permutation_map<(d0) -> (0, 0, 0)> : tensor<1xf32>, vector<4x4x4xf32>
> vector.transfer_read %tensor[%idx, %idx2], permutation_map<(d0, d1) -> (0, 0, 0)> : tensor<1x1xf32>, vector<4x4x
> ```

All these examples involve broadcasting. It’s worth referring to this point from the vector.transfer_read [documentation](https://mlir.llvm.org/docs/Dialects/Vector/#vectortransfer_read-vectortransferreadop) (emphasis mine):

> The mask type is an i1 vector with a shape that matches how elements are read from the MemRef/Tensor, **before** any permutation or **broadcasting**. 

This PR strictly follows the documented design, which ensures consistency. Other parts of MLIR should adhere to the same principle as well.

> Based on the definition, the mask dimensionality should always match the dimensionality of the domain.

Which definition? :)

Using `vector<1x1xf32>` instead of `vector<1xf32>` in your 3rd example would be one possibility, but not the only one ;-)

**Comment by dcaballe** [comment]
Hey Kunwar,

Your feedback is definitely welcome. Currently `vector.mask`, as some other ops, does not support 0-D vectors at ODS level. You can take a look at the differences between:

```
// Vector types.

class VectorOf<list<Type> allowedTypes> :
  ShapedContainerType<allowedTypes, IsVectorTypePred, "vector",
                      "::mlir::VectorType">;

// Temporary vector type clone that allows gradual transition to 0-D vectors.
// TODO: Remove this when all ops support 0-D vectors.
class VectorOfAnyRankOf<list<Type> allowedTypes> :
  ShapedContainerType<allowedTypes, IsVectorOfAnyRankTypePred, "vector",
                      "::mlir::VectorType">;
```

0-D vector support has been a major source of bugs and trouble in general, even for simple ops, so I don't think we can just enable 0-D vectors here and move on. Honestly, I'm really concerned about the 0-D vector situation and the instability that has been bringing since its introduction. For that reason, as I mentioned in your PR, I was thinking about sending an RFC to propose   its removal. It would be good to know if people have any concerns about that beforehand.

**Comment by banach-space** [comment]
@Groverkss , are you OK for Diego to proceed with this? 

It's the first step towards fixing https://github.com/llvm/llvm-project/issues/116197, hence I'm quite keen :) I will be sending a follow-on soonish.

**Comment by Groverkss** [comment]
> @Groverkss , are you OK for Diego to proceed with this?
> 
> It's the first step towards fixing #116197, hence I'm quite keen :) I will be sending a follow-on soonish.

Yes! I will follow up on the discussion elsewhere, this patch by itself is good.

**Pull request merged**